File: VelibStations_Android/app/src/main/java/com/muhammedsafiulazam/velibstations/feature/stationinfo/model/StationInfoModel.kt
package com.muhammedsafiulazam.velibstations.feature.stationinfo.model

import android.content.Context
import android.text.TextUtils
import com.muhammedsafiulazam.common.addon.AddOnManager
import com.muhammedsafiulazam.common.addon.AddOnType
import com.muhammedsafiulazam.common.database.IDatabaseManager
import com.muhammedsafiulazam.common.database.velib.event.VelibDatabaseEventType
import com.muhammedsafiulazam.common.event.Event
import com.muhammedsafiulazam.common.event.IEventManager
import com.muhammedsafiulazam.common.model.Location
import com.muhammedsafiulazam.common.service.IServiceManager
import com.muhammedsafiulazam.common.service.model.Error
import com.muhammedsafiulazam.common.service.velib.event.VelibServiceEventType
import com.muhammedsafiulazam.common.service.velib.model.Dataset
import com.muhammedsafiulazam.common.service.velib.model.Record
import com.muhammedsafiulazam.common.view.BaseModel
import com.muhammedsafiulazam.common.view.IViewManager
import com.muhammedsafiulazam.photoalbum.context.IContextManager
import com.muhammedsafiulazam.velibstations.R
import com.muhammedsafiulazam.velibstations.feature.stationinfo.event.StationInfoEventType
import com.muhammedsafiulazam.velibstations.feature.stationlist.event.StationListEventType

class StationInfoModel : BaseModel() {

    private lateinit var mEventManager: IEventManager
    private var mRecord: Record? = null

    override fun onLoad() {
        super.onLoad()

        // Addons
        mEventManager = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager

        // Enable events.
        receiveEvents(true)
    }

    private fun responseLoadData(response: Any?, error: Error?) {
        val event = Event(StationInfoEventType.MODEL_RESPONSE_LOAD_DATA, response, error)
        mEventManager.send(event)
    }

    override fun onReceiveEvents(event: Event) {
        super.onReceiveEvents(event)
        if (TextUtils.equals(StationInfoEventType.MODEL_REQUEST_LOAD_DATA, event.type)) {
            mRecord = event.data as Record?
            onCreatePropertyList(mRecord)
        }
    }

    fun onCreatePropertyList(record: Record?) {
        try {
            val fields = record?.fields!!
            val propertyList: MutableList<Property> = mutableListOf()

            val contextManager: IContextManager? = AddOnManager.getAddOn(AddOnType.CONTEXT_MANAGEER) as IContextManager?
            val context: Context? = contextManager?.getContext() as Context?

            if (context != null) {

                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_name),
                        fields.name!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_code),
                        fields.code!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_state),
                        fields.state!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_duedate),
                        fields.dueDate!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_latitude),
                        fields.geolocation!!.get(0).toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_longitude),
                        fields.geolocation!!.get(1).toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_kioskstate),
                        fields.kioskState!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_creditcard),
                        fields.creditCard!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_overflowactivation),
                        fields.overflowActivation!!
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_maxbikeoverflow),
                        fields.maxBikeOverflow!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbedock),
                        fields.nbEDock!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbfreeedock),
                        fields.nbFreeEDock!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbdock),
                        fields.nbDock!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbfreedock),
                        fields.nbFreeDock!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbebike),
                        fields.nbEBike!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbbike),
                        fields.nbBike!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbbikeoverflow),
                        fields.nbBikeOverflow!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_nbebikeoverflow),
                        fields.nbEBikeOverflow!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_overflow),
                        fields.overflow!!.toString()
                    )
                )
                propertyList.add(
                    Property(
                        context.getString(R.string.stationinfo_property_densitylevel),
                        fields.densityLevel!!.toString()
                    )
                )
            }

            responseLoadData(propertyList, null)

        } catch (e: Exception) {
            responseLoadData(null, Error(null, e.message))
        }
    }

    override fun onUnload() {
        // Disable events.
        receiveEvents(false)
        super.onUnload()
    }
}

File: VelibStations_iOS/VelibStations/Feature/StationInfo/Model/StationInfoModel.swift
//
//  StationInfoModel.swift
//  VelibStations
//
//  Created by Safi on 20/12/2019.
//

import Foundation
import CommonKit

class StationInfoModel : BaseModel {
    
    private var mRecord: Record? = nil
    private var mEventManager: IEventManager? = nil
    
    override func onLoad() {
        super.onLoad()
        
        mEventManager = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
        
        receiveEvents(receive: true)
    }
    
    private func responseLoadData(response: Any?, error: Error?) {
        let event = Event(type: StationInfoEventType.MODEL_RESPONSE_LOAD_DATA, data: response, error: error)
        mEventManager?.send(event: event)
    }
    
    override func onReceiveEvents(event: Event) {
        super.onReceiveEvents(event: event)
        
        if (StationInfoEventType.MODEL_REQUEST_LOAD_DATA == event.type) {
           mRecord = event.data as? Record
           onCreatePropertyList(record: mRecord)
       }
    }
    
    private func onCreatePropertyList(record: Record?) {
        let fields = record?.fields
        let properties: NSMutableArray = NSMutableArray()
        
        properties.add(Property(title: L10n.StationInfo.Property.name, value: (fields?.name)!))
        properties.add(Property(title: L10n.StationInfo.Property.code, value: (fields?.code)!))
        properties.add(Property(title: L10n.StationInfo.Property.state, value: (fields?.state)!))
        properties.add(Property(title: L10n.StationInfo.Property.Due.date, value: (fields?.dueDate)!))
        
        let latitude = fields?.geolocation?.object(at: 0) as! Double
        let longitude = fields?.geolocation?.object(at: 1) as! Double
        properties.add(Property(title: L10n.StationInfo.Property.latitude, value: "\(latitude)"))
        properties.add(Property(title: L10n.StationInfo.Property.longitude, value: "\(longitude)"))
        
        properties.add(Property(title: L10n.StationInfo.Property.Kiosk.state, value: (fields?.kioskState)!))
        properties.add(Property(title: L10n.StationInfo.Property.Credit.card, value: (fields?.creditCard)!))
        properties.add(Property(title: L10n.StationInfo.Property.Overflow.activation, value: (fields?.overflowActivation)!))
        properties.add(Property(title: L10n.StationInfo.Property.Max.Bike.overflow, value: "\((fields?.kioskState)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.eDock, value: "\((fields?.nbEDock)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.Free.eDock, value: "\((fields?.nbFreeEDock)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.dock, value: "\((fields?.nbDock)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.Free.dock, value: "\((fields?.nbFreeDock)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.eBike, value: "\((fields?.nbEBike)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.bike, value: "\((fields?.nbBike)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.Bike.overflow, value: "\((fields?.nbBikeOverflow)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.EBike.overflow, value: "\((fields?.nbEBikeOverflow)!)"))
        properties.add(Property(title: L10n.StationInfo.Property.Nb.overflow, value: (fields?.overflow)!))
        properties.add(Property(title: L10n.StationInfo.Property.Density.level, value: (fields?.densityLevel)!))
        
        responseLoadData(response: properties, error: nil)
    }
    
    override func onUnload() {
        receiveEvents(receive: false)
        super.onUnload()
    }
}
